metric/gauge/gauge.go

```diff
@@ -0,0 +1,251 @@
+		return entry.(gaugeEntry)
+	}
+}
+
+func (g *Gauge) mapKey(ctx context.Context) gaugeKey {
+	tm := tag.FromContext(ctx)
+	res, ok := resource.FromContext(ctx)
+	if !ok {
+		res = &resource.Resource{}
+	}
+	return gaugeKey{
+		tags: tagsToString(tm),
+		res:  resourceKey{type_: res.Type, labels: resource.EncodeLabels(res.Labels)},
+	}
+}
+
+func tagsToString(tm *tag.Map) string {
+	return string(tag.Encode(tm))
+}
+
+func (g *Gauge) labelValues(s string) []metric.LabelValue {
+	tm, err := tag.Decode([]byte(s))
+	if err != nil {
+		panic(err) // should never happen, since we called Encode
+	}
+	var vals []metric.LabelValue
+	for _, k := range g.keys {
+		val, ok := tm.Value(k)
+		if ok {
+			vals = append(vals, metric.NewLabelValue(val))
+		} else {
+			vals = append(vals, nil)
+		}
+	}
+	return vals
+}
+
+// Float64 represents a float64 value that can go up and down.
+//
+// Float64 maintains a value for each combination of of tag values and resource
+// in the context passed to the Set or Add methods.
+type Float64 struct {
+	Gauge
+}
+
+type float64Entry struct {
+	val uint64
+}
+
+func (e *float64Entry) read(t time.Time) metric.Point {
+	return metric.NewFloat64Point(t, math.Float64frombits(atomic.LoadUint64(&e.val)))
+}
+
+// NewFloat64WithRegistry creates a new gauge with a float64 value.
+func NewFloat64(name, description string, unit metric.Unit, keys ...tag.Key) *Float64 {
+	labelKeys := make([]string, 0, len(keys))
+	for _, k := range keys {
+		labelKeys = append(labelKeys, k.Name())
+	}
+	g := &Float64{
+		Gauge{
+			newEntry: func() gaugeEntry {
+				return new(float64Entry)
+			},
+			keys:  keys,
+			start: time.Now(),
+			desc: metric.Descriptor{
+				Name:        name,
+				Description: description,
+				Unit:        unit,
+				LabelKeys:   labelKeys,
+			},
+		},
+	}
+	return g
+}
+
+// Set sets the current gauge value.
+func (g *Float64) Set(ctx context.Context, val float64) {
+	ge := g.getEntry(ctx).(*float64Entry)
+	atomic.StoreUint64(&ge.val, math.Float64bits(val))
+}
+
+// Add increments the current gauge value by val.
+func (g *Float64) Add(ctx context.Context, val float64) {
+	ge := g.getEntry(ctx).(*float64Entry)
+	var swapped bool
+	for !swapped {
+		oldVal := atomic.LoadUint64(&ge.val)
+		newVal := math.Float64bits(math.Float64frombits(oldVal) + val)
+		swapped = atomic.CompareAndSwapUint64(&ge.val, oldVal, newVal)
+	}
+}
+
+// Int64 represents a int64 gauge value that can go up and down.
+//
+// Int64 maintains a value for each combination of of tag values
+// in the context passed to the Set or Add methods.
+type Int64 struct {
+	Gauge
+}
+
+type int64GaugeValue struct {
+	val int64
+}
+
+func (v *int64GaugeValue) read(t time.Time) metric.Point {
+	return metric.NewInt64Point(t, atomic.LoadInt64(&v.val))
+}
+
+// NewInt64WithRegistry creates a new int64-valued gauge and adds it to the
+// given registry.
+func NewInt64(name, description string, unit metric.Unit, keys ...tag.Key) *Int64 {
```

Comment:
I think keys should be a type of LabelKeys. Please check the Java's implementation. We use TagKey for stats api, where as LabelKey for metrics api.

https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/metrics/MetricRegistry.java#L49

Reply:
Yeah I thought about that, but the tags are what are in the context. Do you think we should not get tags from the context passed in to the Set and Add methods?

Reply:
We accept list of LableValue (length should be same as labelkeys) while adding new entry/point to gauge. Does this answer your question?

Reply:
Updated to use use label values directly.